[future] Avoid some future callback self-captures.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36502 [future] Avoid some future callback self-captures.**

We're sometimes deleting futures without completing them (discovered by logging),
and we've recently noticed a slow memory leak.

This change migrates the future lambda cases where there was self-capture.
 - In some cases, we use weak_ptr<>, plus .lock()/assert in the lambda callback.
   This avoids the reference cycle. We use this primarily in the case where the
   value ends up being moved in the callback (something we want to be careful about)

 - We also add a convenience api to Future where the completed Future is returned as an arg.
   This allows us to avoid self-capture, though it assumes that the markCompleted()
   caller is persisting the future for the markCompleted() duration (this has been the case)

Differential Revision: [D20998905](https://our.internmc.facebook.com/intern/diff/D20998905/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20998905/)!

Commit:
[future] Use weak_ptr<> in some callback self-captures, to avoid ref cycles.

I discovered via logging that sometimes we're deleting futures without
completing them.

This can be problematic from a memory leak perspective, since if the callback
lambda acquires a strong reference to the calling future, it will leak memory -
not just the lambda involved, but also the other strong references.

This change simply switches to capturing a weak pointer instead of a strong
one for the relevant addCallback() futures.

Differential Revision: [D20998905](https://our.internmc.facebook.com/intern/diff/D20998905/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20998905/)!

[ghstack-poisoned]

Commit:
Update on "[future] Use weak_ptr<> in some callback self-captures, to avoid ref cycles."


I discovered via logging that sometimes we're deleting futures without
completing them.

This can be problematic from a memory leak perspective, since if the callback
lambda acquires a strong reference to the calling future, it will leak memory -
not just the lambda involved, but also the other strong references.

This change simply switches to capturing a weak pointer instead of a strong
one for the relevant addCallback() futures.

Differential Revision: [D20998905](https://our.internmc.facebook.com/intern/diff/D20998905/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20998905/)!

[ghstack-poisoned]

jjlilley commented:
fyi: I think this might address the slow memory leak that we were discussing this morning.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -393,7 +398,7 @@ void DistEngine::cleanupBackwardPass(const ContextPtr& autogradContext) {
   const auto& futureGrads =
       autogradContext->retrieveGraphTask()->future_result_;
-  TORCH_INTERNAL_ASSERT(futureGrads.use_count() == 1);
+  TORCH_INTERNAL_ASSERT(futureGrads.use_count() <= 2);
```
This needs to be == 1 and was added so that we don't leak references to the gradients. What is the reason this needs to be changed to <=2?

jjlilley replied:
The 2 is because of the second future reference (since the weak arg becomes strong).
Do you have a preference for a better alternative?

jjlilley replied:
Actually, you might prefer what I send in a few minutes

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -527,7 +527,13 @@ bool ProcessGroupAgent::handleRecv(RecvWork& work) {
       auto fromId = work.from_.id_;
       auto requestId = work.id_;
-      futureResponse->addCallback([this, fromId, requestId, futureResponse]() {
+      futureResponse->addCallback([this,
+                                   fromId,
+                                   requestId,
+                                   weak = std::weak_ptr<FutureMessage>(
+                                       futureResponse)]() {
+        auto futureResponse = weak.lock();
+        TORCH_INTERNAL_ASSERT(futureResponse);
```
Looks like this pattern is repeated a lot. Would it make sense to add an API like `addCallbackCaptureFuture`, where we define a Callback struct consisting of std::function (that takes Future as argument) and a weak ptr to the Future. When the callback is called we can perform weak lock check and pass the std::shared_ptr to the actual callback.

jjlilley replied:
I was thinking that, but was going back-and-forth about whether we wanted that logic in the future class. I might go ahead and do that, might be cleaner

pritamdamania87 reviewed (commented):
Look good, my only concern is the change in the use_count assertion. I agree its not the best way to validate what we want, but we shouldn't change it until we have something better in place.

zhaojuanmao commented:
@jjlilley  are you saying current trunk may have slow memory leak due to this? thanks for fixing it

jjlilley commented:
> @jjlilley are you saying current trunk may have slow memory leak due to this? thanks for fixing it

Yes, Satendra discovered it yesterday during a longer run.

Commit:
Syntactic sugar, avoid ref increment on "[future] Avoid some future callback self-captures."


We're sometimes deleting futures without completing them (discovered by logging),
and we've recently noticed a slow memory leak.

This change migrates the future lambda cases where there was self-capture.
 - In some cases, we use weak_ptr<>, plus .lock()/assert in the lambda callback.
   This avoids the reference cycle. We use this primarily in the case where the
   value ends up being moved in the callback (something we want to be careful about)

 - We also add a convenience api to Future where the completed Future is returned as an arg.
   This allows us to avoid self-capture, though it assumes that the markCompleted()
   caller is persisting the future for the markCompleted() duration (this has been the case)

Differential Revision: [D20998905](https://our.internmc.facebook.com/intern/diff/D20998905/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20998905/)!

[ghstack-poisoned]

jjlilley commented:
This should be ready for a look, memory usage still looks flat with the test. 

A few comments:
 - this change avoids changing the use_count() assert that @pritamdamania87 commented on earlier
 - In a few places, I have the code still using a weak_ptr. It's only a few places, and these are for situations where we're std::move'ing the future, and seems reasonably appropriate.
 - I added a convenience api per @pritamdamania87 's suggestion. Some of this diff honestly isn't strictly necessary for the leak, but rather mechanical moving to that api.

pritamdamania87 commented on torch/csrc/utils/future.h:
```diff
@@ -128,7 +128,11 @@ class TORCH_API Future final {
   }
 
-  private:
+  void addCallback(std::function<void(const Future<T>& future)> cb) {
```
Doesn't have to be part of this PR, but does it make sense to pass a non-const reference to the callback such that we can move it around within the callback? This would avoid the other places where we pass the future in explicitly as a weak_ptr.

jjlilley replied:
btw, I wanted to think about this one (maybe we do)

`const Future<T>&` felt quite safe to do, while `Future<T>&` can be more dangerous (somewhat easier to sabotage oneself) with our pattern of allowing multiple callbacks to be added. 

For step 1, I wanted to address the simpler cases, while effectively "marking" the other cases for future consideration.

pritamdamania87 replied:
Yes, it's true that it is definitely more dangerous. That was the reason in the initial PR I added the assert for use_count instead of moving out the value from the Future (which could break callbacks downstream).

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -200,55 +199,55 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
-        } catch (std::exception& e) {
-          accumulateGradFuture->setErrorIfNeeded(e.what());
+    try {
+      const variable_list& grads = futureGrads.constValue();
```
Just realized that if we move the value out of futureGrads here, we don't need the use_count assert.

pritamdamania87 replied:
cc @wanchaol since this is related to the autograd PR.

wanchaol replied:
Cool! does this means that we don't need to return a share_ptr& for future_results?

pritamdamania87 reviewed (approved):
Looks good! Feel free to address the additional comments in a separate PR since it's better to fix the memory leak soon.

jjlilley commented:
Thanks for looking at this! I'll land as-is, and iterate separately.
